frontend/src/lib/components/Table/utils.ts

```diff
@@ -0,0 +1,215 @@
+import React from "react";
+
+import { convertRemToPixels } from "@lib/utils/screenUnitConversions";
+import { getTextWidthWithFont } from "@lib/utils/textSize";
+
+import type {
+    ColumnGroup,
+    ColGroupDef,
+    DataCellDef,
+    FilterCellDef,
+    HeaderCellDef,
+    TableCellDefinitions,
+    LoadedData,
+    TableColumn,
+    TableData,
+    TableColumns,
+} from "./types";
+
+export function isColumnGroupDef<TData extends Record<string, any>>(
+    headerOrGroup: TableColumn<TData>,
+): headerOrGroup is ColumnGroup<TData> {
+    return "subColumns" in headerOrGroup;
+}
+
+export function isLoadedDataRow<TData extends Record<string, any>>(data: TableData<TData>): data is LoadedData<TData> {
+    if (data._pending) return false;
+    return true;
+}
+
+function recursivelyCalcDepth<TData extends Record<string, any>>(
+    tableColumns: TableColumns<TData>,
+    depth: number = 1,
+): number {
+    let maxDepth = depth;
+    for (const columnDef of tableColumns) {
+        if (isColumnGroupDef(columnDef) && columnDef.subColumns) {
+            const localDepth = recursivelyCalcDepth(columnDef.subColumns, depth + 1);
+            maxDepth = Math.max(maxDepth, localDepth);
+        }
+    }
+    return maxDepth;
+}
+
+export function defaultDataFilterPredicate<TData extends Record<string, any | null>, TKey extends keyof TData>(
+    filterValue: string,
+    columnData: TData[TKey],
+    dataDefinition: DataCellDef<TData, TKey>,
+    entry: TData,
+) {
+    const formattedData = dataDefinition.format?.(columnData, entry) ?? columnData ?? "";
+
+    const lowerValue = String(formattedData).toLowerCase();
+    const filterString = filterValue.toLowerCase();
+
+    return lowerValue.includes(filterString);
+}
+
+export function recursivelyBuildTableCellDefinitions<TData extends Record<string, any>>(
+    tableColumns: TableColumns<TData>,
+    depth: number = 0,
+    maxDepth = recursivelyCalcDepth(tableColumns),
+    // ! Object is mutated as the method runs
+    headerCells: HeaderCellDef[][] = [],
+): TableCellDefinitions<TData> {
```

Comment:
Consider creating an exported version with only `tableColumns`-prop. And have a private (non-exported) version which has the headerCells mutated?

Seems like we're only using it in `table-tsx`, and then only using `tableColumns`-prop.